File: 11-flyweight/11-flyweight.go
package main

import "fmt"

type Character struct {
	char string
}

func (c Character) render() {
	fmt.Println("Rendering:", c.char)
}

type CharacterFactory struct {
	chars map[string]Character
}

func (cf *CharacterFactory) getChar(c string) Character {
	if _, ok := cf.chars[c]; !ok {
		cf.chars[c] = Character{char: c}
	}
	return cf.chars[c]
}

func (cf *CharacterFactory) getCharsAndLen() {
	fmt.Printf("Total unique characters: %d\n", len(cf.chars))
	for _, v := range cf.chars {
		fmt.Printf(" - %s\n", v.char)
	}
}

func main() {
	factory := &CharacterFactory{chars: make(map[string]Character)}

	word := "Hello World"
	for _, c := range word {
		if c != ' ' {
			char := factory.getChar(string(c))
			char.render()
		}
	}

	factory.getCharsAndLen()
}

File: 11-flyweight/11-flyweight.java
import java.util.HashMap;
import java.util.Map;

interface Character {
    String render();
}

class ConcreteCharacter implements Character {
    private final String ch;

    public ConcreteCharacter(String ch) {
        this.ch = ch;
    }

    @Override
    public String render() {
        return this.ch;
    }
}

class CharacterFactory {
    private final Map<String, ConcreteCharacter> chars = new HashMap<>();

    public Character getChar(String ch) {
        if (!this.chars.containsKey(ch)) {
            this.chars.put(ch, new ConcreteCharacter(ch));
        }

        return this.chars.get(ch);
    }

    public int getCacheSize() {
        return this.chars.size();
    }
}

public class Main {
    public static void main(String[] args) {
        CharacterFactory factory = new CharacterFactory();

        Character c1 = factory.getChar("A");
        Character c2 = factory.getChar("B");
        Character c3 = factory.getChar("A");

        System.out.println(c1.render()); // A
        System.out.println(c2.render()); // B
        System.out.println(c3.render()); // A

        System.out.println("Total unique chars in memory: " + factory.getCacheSize()); // 2
    }
}

